src/lexer.rs

```diff
@@ -17,53 +17,55 @@ type CharAndLocation = (char, Location);
 //
 // It expects that phases 1 to 3 are already done and that valid utf8 is passed in.
 #[derive(Clone)]
-pub struct CharsAndLocation<'a> {
-    chars: Peekable<Chars<'a>>,
-    loc: Location,
+pub struct CharsAndLine<'a> {
+    inner: Chars<'a>,
+    line: u32,
 }
 
-impl<'a> CharsAndLocation<'a> {
+impl<'a> CharsAndLine<'a> {
     pub fn new(input: &'a str) -> Self {
-        CharsAndLocation {
-            chars: input.chars().peekable(),
-            loc: Location::default(),
+        CharsAndLine {
+            inner: input.chars(),
+            line: 1,
         }
     }
+
+    pub fn get_current_ptr(&self) -> *const u8 {
+        self.inner.as_str().as_bytes().as_ptr()
+    }
 }
 
-impl<'a> Iterator for CharsAndLocation<'a> {
-    type Item = CharAndLocation;
+impl<'a> Iterator for CharsAndLine<'a> {
+    type Item = CharAndLine;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let current = self.chars.next()?;
-        self.loc.start = self.loc.end;
-        self.loc.end += current.len_utf8() as u32;
+        let current = self.inner.next()?;
 
         match current {
             '\n' => {
                 // Consume the token but see if we can grab a \r that follows
-                if self.chars.peek() == Some(&'\r') {
-                    self.chars.next();
-                    self.loc.end += 1;
+                let mut peek_inner = self.inner.clone();
+                if peek_inner.next() == Some('\r') {
+                    self.inner = peek_inner;
```

Comment:
This was going to be the controversial change in this PR :)

The problem is that a peekable for `CharsAndLine` forces peekable for `SkipComents` and `LexerInnerIterator` too (because Peekable is not Cloneable). My concern with using a lot of chained peekables is that it adds a bunch of control flow for each `LexerInnerIterator.next()` that's not needed because our iterators are almost stateless.

If you want I can try to undo this change, and put it in a separate PR so we can discuss there.

Reply:
I still don't quite know how the change that makes LexerToken contain a CoW will look like, but it's very likely that it will make the iterators even more stateless because we likely need to pass the whole `&str` from the outside of the iterators (which wouldn't be iterators anymore).

Reply:
I don't think we need to undo this change I just find `Peekable` more idiomatic but it's hard to measure performance here without running a proper benchmark since there are things like cache locality which shouldn't really affect this too much but you never know. I think we can merge this has is and in the future if we feel like it look at performance here

Reply:
Thank you.